File: src/main/py/weather.py
import requests # pip install requests
import json
from pprint import pprint

# https://home.openweathermap.org/api_keys
# api key: 04afad80f7b95195a3a9e49fa1659949  // for amit r.
# https://openweathermap.org/current
# https://api.openweathermap.org/data/2.5/weather?lat={lat}&lon={lon}&appid={API key}
# http://api.openweathermap.org/geo/1.0/direct?q={city name},{state code},{country code}&limit={limit}&appid={API key}
# http://api.openweathermap.org/geo/1.0/zip?zip={zip code},{country code}&appid={API key}
# https://www.iso.org/obp/ui/#search
# $HOMEBREW_PREFIX/opt/python@3.13/libexec/bin

BASE_URL = "https://api.openweathermap.org/"
API_KEY = "04afad80f7b95195a3a9e49fa1659949" 


def get_city_coords(zip):
    params = "geo/1.0/zip?zip={zip},US"
    url = BASE_URL + params + API_KEY
    response = requests.get(url)
    data = json.loads(response.text)
    return data

def get_weather_by_coords(zip):
    coords = get_city_coords(zip)
    lat = coords[0]["lat"]
    lon = coords[0]["lon"]

    params = "data/2.5/weather?lat={lat}&lon={lon}"
    url = BASE_URL + params + API_KEY
    response = requests.get(url)
    data = json.loads(response.text)
    return data

def ask_zip_show_weather():
    zip = input("Enter zip code: ")
    data = get_weather_by_coords(zip)
    pprint(data)


if __name__ == "__main__":
    print(ask_zip_show_weather())    

File: src/main/java/weather.java

import java.io.FileInputStream;
import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.Base64;
import java.util.List;
import java.util.Properties;
import java.util.Scanner;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;

public class weather {
    private static final String BASE_URL = "https://api.openweathermap.org/";
    private static final String API_PARAM_STRING = "&appid=";

    public static void main(String[] args) {
        askZipShowWeather();
    }

    public static void askZipShowWeather() {
        try (Scanner kbReader = new Scanner(System.in)) {
            System.out.println("Enter zip code: ");
            String zip = kbReader.next();
            String data = getWeatherByZipcode(zip);
            System.out.println("Lets check Weather for " + zip + ":");
            System.out.println(data);
        }
    }

    private static String getWeatherByZipcode(String zip) {
        List<String> coords = getCityCoordinates(zip);
        String lat = coords.get(0);
        String lon = coords.get(1);
        String data = getWeatherByCoords(lat, lon);
        return data;
    }

    private static String getWeatherByCoords(String lat, String lon) {
        String params = String.format( "data/2.5/weather?lat=%s&lon=%s", lat, lon);
        String url = BASE_URL + params + getApiKey();
        HttpRequest request = HttpRequest.newBuilder().uri(URI.create(url)).build();
        HttpClient client = HttpClient.newHttpClient();
        HttpResponse<String> response = null;
        try {
            response = client.send(request, HttpResponse.BodyHandlers.ofString());
        } catch (Exception e) {
            throw new RuntimeException(e);
        }

        if (response != null && response.statusCode() != 200) {
            throw new RuntimeException("Failed : HTTP error code : " + response.statusCode());
        }
        String data = response.body();
        JsonObject obj = JsonParser.parseString(data).getAsJsonObject();
        Gson gson = new GsonBuilder().setPrettyPrinting().create();
        data = gson.toJson(obj);
        return data;
    }

    private static List<String> getCityCoordinates(String zip){
        String params = String.format("geo/1.0/zip?zip=%s,US", zip) ;
        String url = BASE_URL + params + getApiKey();
        HttpRequest request = HttpRequest.newBuilder().uri(URI.create(url)).build();
        HttpClient client = HttpClient.newHttpClient();
        HttpResponse<String> response = null;
        try {
            response = client.send(request, HttpResponse.BodyHandlers.ofString());
        } catch (Exception e) {
            throw new RuntimeException(e);
        }

        if (response != null && response.statusCode() != 200) {
            throw new RuntimeException("Failed : HTTP error code : " + response.statusCode());
        }
        String data = response.body();
        JsonObject obj = JsonParser.parseString(data).getAsJsonObject();
        
        List<String> coords = new ArrayList<>();
        coords.add(obj.get("lat").toString());
        coords.add(obj.get("lon").toString());
        return coords;
    }

    private static String getApiKey() {
        Path path = Paths.get("src/main/resources/application.properties");
        StringBuffer apiKey = new StringBuffer("&appid=");
        try {
            FileInputStream fis = new FileInputStream(path.toFile());
            Properties appProperties = new Properties();
            appProperties.load(fis);
            String encodedKey = appProperties.getProperty("apiKey");
            String key = Base64.getDecoder().decode(encodedKey).toString();

            String key2 = appProperties.getProperty("open.weather.api.key");

            apiKey.append(key2);

        } catch (IOException e) {
            System.out.println(e.getMessage());
        }
        return apiKey.toString();
    }
}
